sdk/cognitiveservices/azure-cognitiveservices-formrecognizer/DESIGN_DOC.md

```diff
@@ -0,0 +1,544 @@
+# Form Recognizer Design Overview
+
+The Form Recognizer client library provides two clients to interact with the service: `FormRecognizerClient` and 
+`CustomFormClient`, which can be imported from the `azure.ai.formrecognizer` namespace. The asynchronous clients 
+can be imported from `azure.ai.formrecognizer.aio`.
+
+`FormRecognizerClient` provides methods for interacting with the receipt and layout models.
+`CustomFormClient` provides the methods for training custom models to analyze forms.
+
+Authentication is achieved by passing an instance of `CognitiveKeyCredential("<api_key>")` to the client,
+or by providing a token credential from `azure.identity` to use Azure Active Directory.
+
+## Receipt and Layout
+
+Receipt and layout are accessed through the `FormRecognizerClient`. The input form or document can be passed as a 
+string url/path to the image, or as a file stream. The SDK will determine content-type and send the appropriate header. 
+
+Both receipt and layout methods return poller objects which are used to get the result.
+The `begin_extract_receipt` method returns a `List[ExtractedReceipt]` with hardcoded receipt fields.
+The `begin_extract_layout` method returns the extracted layouts as a `List[ExtractedLayoutPage]`.
+
+If the keyword argument `include_text_details=True` is passed in, the `elements` attribute will be re-hydrated with the
+OCR result for the particular value/cell referenced by the json pointer.
+
+### Form Recognizer Client
+```python
+from azure.ai.formrecognizer import FormRecognizerClient
+
+client = FormRecognizerClient(endpoint: str, credential: Union[CognitiveKeyCredential, TokenCredential])
+
+client.begin_extract_receipt(form: Union[str, BytesIO], **kwargs) -> LROPoller -> List[ExtractedReceipt]
+
+client.begin_extract_layout(form: Union[str, BytesIO], **kwargs) -> LROPoller -> List[ExtractedLayoutPage]
+```
+
+### Receipt Models
+```python
+class ExtractedReceipt(DictMixin):
+    receipt_items: List[ReceiptItem]
+    merchant_address: FieldValue
+    merchant_name: FieldValue
+    merchant_phone_number: FieldValue
+    receipt_type: FieldValue
+    subtotal: FieldValue
+    tax: FieldValue
+    tip: FieldValue
+    total: FieldValue
+    transaction_date: FieldValue
+    transaction_time: FieldValue
+    page_range: PageRange
+    page_metadata: PageMetadata
+
+class ReceiptItem:
+    name: FieldValue
+    quantity: FieldValue
+    item_price: FieldValue
+    total_price: FieldValue
+
+class FieldValue:
```

Comment:
Coming back to generics. I know it's just a type hint, but we can make `value` explicitly typed. This would replace the Union with `T`.

Reply:
Agreed. We'll sort out the type annotations in the actual PR. The high-order bits are 1) generated reference documentation has to look good (both docfx and sphinx generated code), 2) no runtime performance hits (which can happen if you are not careful in how you define you generic types) and 3) we don't accidentally over document the generics (which causes unnecessary breaking changes down the line).